sdks/java/io/tika/src/main/java/org/apache/beam/sdk/io/tika/TikaIO.java

```diff
@@ -262,49 +161,50 @@ public void populateDisplayData(DisplayData.Builder builder) {
       }
       Metadata metadata = getInputMetadata();
       if (metadata != null) {
-        StringBuilder sb = new StringBuilder();
-        sb.append('[');
-        for (String name : metadata.names()) {
-            if (sb.length() > 1) {
-              sb.append(',');
-            }
-            sb.append(name).append('=').append(metadata.get(name));
-        }
-        sb.append(']');
+        //TODO: use metadata.toString() only without a trim() once Apache Tika 1.17 gets released
         builder
-            .add(DisplayData.item("inputMetadata", sb.toString())
+            .add(DisplayData.item("inputMetadata", metadata.toString().trim())
             .withLabel("Input Metadata"));
       }
-      if (Boolean.TRUE.equals(getParseSynchronously())) {
-        builder
-          .add(DisplayData.item("parseMode", "synchronous")
-            .withLabel("Parse Mode"));
-      } else {
-        builder
-          .add(DisplayData.item("parseMode", "asynchronous")
-            .withLabel("Parse Mode"));
-        builder
-          .add(DisplayData.item("queuePollTime", getQueuePollTime().toString())
-            .withLabel("Queue Poll Time"))
-        .add(DisplayData.item("queueMaxPollTime", getQueueMaxPollTime().toString())
-          .withLabel("Queue Max Poll Time"));
-      }
-      Integer minTextLen = getMinimumTextLength();
-      if (minTextLen != null && minTextLen > 0) {
-        builder
-        .add(DisplayData.item("minTextLen", getMinimumTextLength().toString())
-          .withLabel("Minimum Text Length"));
-      }
-      if (Boolean.TRUE.equals(getReadOutputMetadata())) {
-        builder
-          .add(DisplayData.item("readOutputMetadata", "true")
-            .withLabel("Read Output Metadata"));
-      }
     }
 
-    @Override
-    protected Coder<String> getDefaultOutputCoder() {
-      return StringUtf8Coder.of();
+    private static class ParseToStringFn extends DoFn<ReadableFile, ParseResult> {
+
+      private static final long serialVersionUID = 6837207505313720989L;
+      private final TikaIO.ParseAll spec;
+      ParseToStringFn(TikaIO.ParseAll spec) {
+        this.spec = spec;
+      }
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        ReadableFile file = c.element();
+        InputStream stream = Channels.newInputStream(file.open());
+        try (InputStream tikaStream = TikaInputStream.get(stream)) {
+
+          TikaConfig tikaConfig = null;
+          if (spec.getTikaConfigPath() != null) {
```

Comment:
It might be a good idea to do this in `@Setup` rather than on every element.

Reply:
Do you suggest creating TikaConfig in the ParseToStringFn constructor or even earlier, in ParseAll, in withTikaConfigPath ?

Reply:
I'm suggesting to do this in the `@Setup` method of this `DoFn`, which is invoked once for every created instance of this DoFn (there'll of course be many instances; at least 1 per worker per thread). It's commonly used to establish connections, cache configuration etc. See e.g. https://github.com/apache/beam/blob/master/sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java#L473

This shouldn't be done in constructor or builder method, because those methods are invoked when the pipeline is constructed, rather than when it is executed. When a user specifies a path to the config, they probably expect that the config will be read at execution time, rather than at construction time. This matters really a lot of if we're talking about a template pipeline, which is constructed once but may be executed many times.

I suppose it might make sense to allow the user to pass an explicitly specified config at construction time (as a Java string containing the XML), if they don't want to bother copying the Tika config onto a distributed filesystem just so it's accessible to the pipeline and if they prefer instead to inline it into the pipeline. You might want to add this feature.

Reply:
I see, thanks, did not know about DoFn possibly having setup/teardown, just updated... I'm planning to have another PR at the next stage which will address all of the Configuration relates improvements (custom content handlers, etc, and now including the possibility of passing the XML configuration fragment as you suggested). 
Re the shortcut and ParseResult success/failure, I've np with continuing looking into it in this PR, but may be it will be easier, esp for the reviewers, to merge what is already available, this IO is still Experimental so I guess it will be safe enough, but it's up to the team